Specify MeterProvider configurable cardinality limits (#2960)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

    `exemplar_reservoir` (optional) to use for storing exemplars. This should be
    a factory or callback similar to aggregation which allows different
    reservoirs to be chosen by the aggregation.
  * **Status**: [Experimental](../document-status.md) - the
    `cardinality_limit` (optional) associated with the view.  This
    should be an positive integer to be taken as a hard limit on the
    number of data points that will be emitted during a single
    collection by a single instrument.  See [cardinality limits](#cardinality-limits),

>     below.

Reviewer: You've removed the meterprovider level configuration in favor of only a view level option. While I think its ok to have a view level cardinality limit option, it shouldn't be the only way because views have sharp edges that make them somewhat surprising to use:

Suppose I want to change the default cardinality limit for all instruments.. I register a view that selects all instruments and sets a cardinality limit of let's say 100. Separately, I want to change the histogram buckets boundaries for an instrument, so I register a view which selects `http.server.duration` and sets the bucket boundaries to be my new values, let's say `[1, 2, 3, 4, 5]`. 

A user is might expect to have `http.server.duration` output with a cardinality limit of 100, and bucket boundaries of `[1, 2, 3, 4, 5]`, but in reality they'll get two metrics:
- One with name `http.server.duration`, a cardinality limit of 100, and the default bucket boundaries of `[0, 5, 10, ...]`
- Another with name `http.server.duration`, the default cardinality limit of 2000, and bucket boundaries of `[1, 2, 3, 4, 5]`. 

This somewhat surprising behavior falls out from the fact that we don't try to do any merging when multiple views match an instrument. Users must be careful that each instrument has only one matching view or get confusing duplicated metrics.

This is a good enough reason to add a meterprovider level configuration option.

Author: I'm interested in breaking out a separate issue to discuss the problem brought up by your example above.

The way I was thinking about it, the View mechanism could configure the cardinality limit, but like with other defaults, there would be a per-instrument-kind default specified at the `MetricReader` level. I've added one bullet to the constructor for `MetricReader` to reflect this position.

Reviewer 2: > This is a good enough reason to add a meterprovider level configuration option.

+1

    `exemplar_reservoir` (optional) to use for storing exemplars. This should be
    a factory or callback similar to aggregation which allows different
    reservoirs to be chosen by the aggregation.
  * **Status**: [Experimental](../document-status.md) - the
    `aggregation_cardinality_limit` (optional) associated with the view.  This

>     should be an positive integer to be taken as a hard limit on the

Reviewer 3: ```suggestion
    should be a positive integer to be taken as a hard limit on the
```

The implementation MUST complete the execution of all callbacks for a
given instrument before starting a subsequent round of collection.

### Cardinality limits

**Status**: [Experimental](../document-status.md)

Views SHOULD support being configured with a cardinality limit to be
applied to all aggregators not configured by a specific view.

View configuration SHOULD support applying per-view cardinality linits.

The cardinality limit is taken as an exact, hard limit on the number
of data points that can be written per collection.  Each view MUST NOT
output more than the configured cardinality limit number of data ponts
per period.

An overflow attribute is defined, named `otel.metric.overflow` having

> (boolean) value `true`, used to report a synthetic aggregation of the

Reviewer 4: `otel.metric.overflow` would be `true` until the process restarts, right? There is no way it is ever set to `false`, or am I missing something?

Reviewer: If temporality is delta, all the series are cleared on collection and the `otel.metric.overflow=true` series goes away with the rest of them. But yet, `otel.metric.overflow` should never be `false`.

Author: I clarified this topic, hopefully it's better now.

metric events that could not be independently aggregated because of

> the limit.  The overflow attribute set MUST be included in the limit

Reviewer 5: I read this multiple times and might be just me being stupid but I lack a clear, normative statement of what SDKs should do. I feel that the text is implicitly telling that SDKs need to put all data into a new stream that contains ONLY the overflow attribute but I wonder if it wouldn't be best to "split" the explanation of the problem and WHAT the SDK should do. Something like:

>An overflow attribute is defined, named `otel.metric.overflow`. The overflow attribute is used to report a synthetic >aggregation of the metric events that could not be independently aggregated because of the limit. 

> The overflow attribute set MUST be included in the limit calculation, thus the maximum number of distinct, non-overflow attributes is one less than the limit.

> When the cardinality limit is reached, the SDK MUST create a new metric stream containing the `otel.metric.overflow`. The attribute MUST have a value of `true`

Author: Reworded:
```
An overflow attribute set is defined, containing a single attribute
`otel.metric.overflow` having (boolean) value `true`, which is used to
report a synthetic aggregation of the metric events that could not be
independently aggregated because of the limit.

The SDK MUST create an Aggregator with the overflow attribute set
prior to reaching the cardinality limit and use it to aggregate events
for which the correct Aggregator could not be created.  The maximum
number of distinct, non-overflow attributes is one less than the
limit, as a result.
```
How's that?

Reviewer 5: Looks great!

calculation, thus the maximum number of distinct, non-overflow

> attributes is one less than the limit.

Reviewer 4: We only need to add 1 here since adding the `otel.metric.overflow=true` does not change the number of metric streams that are open. In other words, adding the same attribute to all streams does not make the cardinality explode, correct?

Reviewer 4: Or would, as soon as the overflow is detected, all further data collected on the metrics stream that has no attributes **but** the otel.metric.overflow attribute?

Reviewer: > Or would, as soon as the overflow is detected, all further data collected on the metrics stream that has no attributes but the otel.metric.overflow attribute?

This.

Author: I reworded this, PTAL.

The RECOMMENDED default cardinality limit is 2000.

#### Synchronous instrument cardinality limits

Views of synchronous instruments with cumulative aggregation
temporality MUST continue export the all attribute sets that were
observed prior to the beginning of overflow.  Metric events
corresponding with attribute sets that were not observed prior to the
overflow will be reflected in a single data point described by (only)

> the overflow attribute.

Reviewer 4: Will the single data point contain the sum of all the other metrics streams that we were unable to keep in memory since the cardinality limit was reached?

Reviewer: Yes.

Reviewer 4: Ah, yeah, that makes sense, I didn't think about that. Thanks for clarifying!

Views of synchronous instruments with delta aggregation temporality

> can choose an arbitrary subset of attributes to output.  Regardless of

Reviewer 4: An arbitrary subset of attribute sets, that is, correct? An attribute set would represent one metrics stream in memory.

Author: This sentence was misleading. I've reworded the section to read:
```
Views of synchronous instruments with delta aggregation temporality
MAY choose an arbitrary subset of attributes to output.

Regardless of aggregation temporality, the SDK MUST ensure that every
metric event is reflected in exactly one Aggregator, which is either
an Aggregator associated with the correct attribute set or an
aggregator associated with the overflow attribute set.

Events MUST NOT be double-counted or dropped during overflow an
overflow.
```

The implementation MUST complete the execution of all callbacks for a
given instrument before starting a subsequent round of collection.

> ### Cardinality limits

Reviewer 6: Maybe I'm misunderstanding, but it seems like this cardinality limit only protects the metric pipeline that the data is being sent to, but does not help prevent memory abuse due to cardinality problems on the process collecting the metrics in the first place.  This is unfortunate as that is usually something a user cares strongly about. Users should be given the tools to put cardinality limits in place for each instrument, in addition to being able to configure default limits for all instruments.

Reviewer 6: Is the cardinality limit on ingest/collection in the process, or only on export? If the limit is on the in-memory state, then it works for both.  If the limit is only for export, then the in-memory representation can still grow unconstrained.

Reviewer 6: Ok, sounds good.  Thanks for clarifying.

Author: https://github.com/open-telemetry/opentelemetry-specification/pull/2960#issuecomment-1515244723

time. Attribute truncation or deletion could affect identity of metric time
series and the topic requires further analysis.

## Cardinality limits

**Status**: [Experimental](../document-status.md)

MeterProviders SHOULD support being configured with limits to be applied
to metrics produced from a single Instrumentation Library Scope.  This
mechanism supports protecting metrics pipelines from excessive da
ta

> production, which is usually caused by a problem known as ["explosition

Reviewer 2: I think explosion is the intended word.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/supplementary-guidelines.md#synchronous-example-cumulative-aggregation-temporality

_Imagine if we have a long running service and we collect metrics with 7 attributes and each attribute can have 30 different values. We might eventually end up having to remember the complete set of all 21,870,000,000 permutations! This cardinality explosion is a well-known challenge in the metrics space._

of cardinality"](supplementary-guidelines.md#synchronous-example-cumulative-aggregation-temporality),
in which the number of timeseries produced by application code exceeds a

> reasonable number.  These limits help protect the overall system from

Author: I now think it would be best to avoid the new term, and simply say what we mean. I've written:

```
MeterProviders SHOULD support being configured with limits to be
applied to metrics produced from a single metric instrument.  This
mechanism supports protecting metrics pipelines from excessive data
production in cases when the number of timeseries produced by
application code grows large, which can happen due to several factors.

Whether because of single attributes having many distinct values or
because of combinatorial expansion among many attributes, these limits
help protect the overall system from individual sources of excessive
metrics instrumentation.
```

individual sources of excessive metrics instrumentation.

When the limits are reached by an the individual instrumentation
library scope, the entire batch of metrics for the scope MUST be

> dropped and an error reported to the user indicating:

Reviewer 7: I interpret this as saying if limits are reached than no data will be exported for that instrumentation scope that collection cycle. Is this the intention?

If so, it seems like we could fail "more gracefully" by dropping the data that exceeded the limit and exporting the previously recorded data (while still emitting an error).

Reviewer: > If so, it seems like we could fail "more gracefully" by dropping the data that exceeded the limit and exporting the previously recorded data (while still emitting an error).

In some cases that would be useful, but in other cases that would be harmful as the continued flow of some data might reduce suspicion that something is wrong. If I don't have reason to suspect something's wrong, I don't check the logs and don't see the error.

Reviewer 2: One approach I've seen in Microsoft - have a "catch-all" stream, for example, if we have the following streams:
* x="a", y="a", sum=3
* x="a", y="b", sum=4
* x="b", y="a", sum=5
* x="b", y="b", sum=6
* x="c", y="a", sum=10
* x="c", y="b", sum=10
* x="c", y="c", sum=10

We can cap it at 5 and report:
* x="a", y="a", sum=3
* x="a", y="b", sum=4
* x="b", y="a", sum=5
* x="b", y="b", sum=6
* x="DIM_CAP", y="DIM_CAP", sum=30

Reviewer 2: I feel "MUST" is too strong, maybe start with "SHOULD" so later we can adjust the behavior (e.g. making it configurable).

Reviewer: That's a nice trick!

Author: @reyang 

> One approach I've seen in Microsoft - have a "catch-all" stream, for example, if we have the following streams

A similar (better, IMO) approach would use unbiased weighted sampling to select 5 of the 7 streams (in your example), meaning to adjust the sum-or-rate-value of sampled timeseries. This way, averaged over time, the expected value of every timeseries will be correct. This is not the time or place to try and formalize this notion, but [it really can be done (e.g., here's a library that does it)](https://github.com/lightstep/varopt) and this is likely when I would configure unlimited cardinality for the MeterProvider and MetricReader. In my exporter, I would apply the downsampling from over-limit cardinality to stay inside of a configured limit.

Reviewer 2: @jmacd thanks for sharing! One downside I can see is the lack of high-fidelity total sum (e.g. with the "DIM_CAP" example above, we have high confidence that the total sum is `3+4+5+6+30`).

It seems based on discussions here we should give some flexibility instead of forcing the data to be dropped?

Reviewer 8: Dropping the data silently will negatively impact service developers. This leads to the drop of data result in firing incidents and the engineers have no clue what's going wrong and spend hours looking into wrong direction before they can finally realize it's drop of data. There needs to be a way to actively signal the consumers of the SDK to know when the data is bring dropped due to hitting cardinality limit.

There definitely needs to be some hook or some sort of mechanism to provide this indication reliably to the customers.

Reviewer 8: We have already run into similar problems where if the data is not dropped the memory grows unbounded and if it's dropped silently then the metrics are wrong indicating problems with the service while the actual issue is in the telemetry.

Author: @dpk83 What would you suggest the SDK can do when the limit is reached for an instrument? It will produce an error and go through standard error handlers, I think.

For my vendor, I am encouraging the use of "lowmemory" temporality which allows delta temporality instruments to be dropped when they are not in use, however as @pirgeo points out, we haven't really defined what "stale" means.

Reviewer 8: > @dpk83 What would you suggest the SDK can do when the limit is reached for an instrument? It will produce an error and go through standard error handlers, I think.

SDK can emit a different metric when events are dropped/throttled. Developers can setup alerts on the metrics to monitor whenever any metrics are throttled/dropped, so it's not lost silently. 
Another mechanism is to give the developers a hook to get a callback when metrics are dropped. Developers can then themselves emit another metrics/log to monitor these cases.

Author: @dpk83 I support this notion, but I do not think that concern should block this PR.  If a user would prefer the risk of running out of memory to dropping metrics, they just need to configure a limit that is so large they would OOM before metrics are dropped.

I think we as a community are interested in someone picking up this project, to prototype and propose the specification for how to count dropped metrics/spans/logs. We have done some preparatory work in this regard, with the introduction of `PartialSuccess` fields.

Author: @dpk83 See https://github.com/open-telemetry/opentelemetry-specification/issues/1655

Reviewer 8: @jmacd  I don't intend to block this PR, as the topic is related so I wanted to make sure that this is considered and there is an item to address the problem to provide the right experience.

> If a user would prefer the risk of running out of memory to dropping metrics, they just need to configure a limit that is so large they would OOM before metrics are dropped.

Regarding the above, the main issue here is the fact that a lot of times users end up adding a high cardinality dimension without either being aware of the fact that the dimension is high cardinality or being aware of the impact of this on memory. In this case the users don't really want the memory to grow unbounded but when such a thing happens they need to be able to catch/spot such problems so that they can take appropriate action and fix it. 

With the silent dropping of the metrics when they go beyond the cardinality we are just going to make the life of these users difficult. Providing a mechanism to easily catch such issues and get an indication quickly is very valuable. Think of this just like a throttling case for any scenarios, if the events are throttled the effect of throttling will show up in a different form to the user and without any throttling metrics user will end up spending a lot of time investigating all parts which could result in the issue.

Author: I marked this conversation resolved because I do not believe the benefit of some kind of graceful degradation when the cardinality limit is reached is worth the amount of code and complexity that will go into it.

I also think we have different definitions of "silent". The user would see an error log message when the limit is reached. Besides, and orthogonal to this issue, [users should be counting dropped metric data points](https://github.com/open-telemetry/opentelemetry-specification/issues/1655). I think our time and energy will be better spent (1) discussing how to count dropped metric data points, spans and logs generally, (2) discussing how to make SDK logs more actionable/monitorable so that users do not experience any kind of silent telemetry outage.

(Besides, if we are prioritizing how to focus our attention on the metrics specification, I believe we have agreed that https://github.com/open-telemetry/opentelemetry-specification/issues/2923 is more important.)

Author: PTAL latest changes.

- Detail about the limit that was exceeded (e.g., the library name and
  version, the configured limit, the instruments of that library
  ranked by cardinality).
- Recommend the user to configure a [View](#view) that filters the
  offending cardinality or drops selected instruments to correct the problem.

Note that limits expressed here refer to exported data for a single
collection cycle.  The number of timeseries referred to in each of
these limits equals the number of distinct data points produced over
one interval.

Under some circumstances, depending on the configured aggregation
temporality, the MeterProvider may be able to recover after these
limits are reached.  In particular, for Synchronous instruments
configured with Delta aggregation temporality, the MeterProvider

> SHOULD remove timeseries from memory when they become stale, making it

Reviewer 4: Do we have a definition for stale somewhere?

Author: I would like us to have a definition somewhere that helps explain how delta temporality allows SDKs to drop memory, but I realize that this PR probably isn't the place to do it. I believe we should add something in the data model to say when a point can be omitted by an exporter, then by extension if the SDK internal state allows removing memory that would produce a point that can be omitted by an exporter, it is allowed to do so.

possible to recover from an explosion of cardinality from one
collection cycle to the next provided the application stops using
excessive cardinality.  However, for MeterProviders that have
Synchronous instruments configured with Cumulative aggregation
temporality, once a limit is reached the MeterProvider SHOULD disable
it for the remaining lifetime of the MeterProvider.

The configurable limits are:

1. The number of timeseries per instrument, with recommended default value 2000.
2. The number of timeseries per instrumentation library scope, with recommended default value 2000.

For both of these limits, the limit MAY be unset to indicate an

> unlimited configuration and the zero value is considered invalid.

Reviewer 4: Would it make sense to treat 0 (or everything less than or equal to 0) as unset as well?

Reviewer 2: A stepping back question - why do we want to have unlimited configuration?

Reviewer 4: Good point. I guess if you don't know what you can expect, but you regularly exceed the 2000 TS limit? You could set it to 0 and run it for a while to see if it stabilized somewhere? Not advocating for it, I think having sensible defaults is good.

Reviewer 2: > You could set it to 0 and run it for a while to see if it stabilized somewhere?

If the goal is to get some rough idea about the cardinality, maybe setting something really large (e.g. 10 million)?

Author: Agreed. I'm happy to remove the "unlimited" configuration as an option and recommend very large limits.

Author: This is done, thanks.

time. Attribute truncation or deletion could affect identity of metric time
series and the topic requires further analysis.

## Cardinality limits

**Status**: [Experimental](../document-status.md)

MeterProviders SHOULD support being configured with limits to be
applied to metrics produced from a single metric instrument.  This
mechanism supports protecting metrics pipelines from excessive data
production in cases when the number of timeseries produced by
application code grows large, which can happen due to several factors.

Whether because of single attributes having many distinct values or
because of combinatorial expansion among many attributes, these limits
help protect the overall system from individual sources of excessive
metrics instrumentation.

When the limit is reached by an the individual metric instrument, the

> entire batch of metrics for the instrument MUST be dropped and an

Reviewer 7: I thought the resolution of [this comment](https://github.com/open-telemetry/opentelemetry-specification/pull/2960#discussion_r1035295900) meant there would be a more nuanced truncation strategy here. Like all additional timeseries being flattened into a single one?

Did I miss something?

Author: Thanks. I followed up in https://github.com/open-telemetry/opentelemetry-specification/pull/2960#discussion_r1063803281. I don't think the effort and complexity of a more nuanced strategy is worthwhile.

Reviewer 7: Notes from the is morning's spec SIG meeting:

> Consolidating all instrument measurement values into a single "OVERFLOW" attribute when cardinality exceeds the configured limit seems to have consensus on the approach here.

Author: PTAL latest changes.

error reported to the user indicating:

- Detail about the limit that was exceeded (e.g., the library name and
  version, the instrument, and the configured limit).
- Recommend the user to configure a [View](#view) that filters the
  offending instrument to limit cardinality or otherwise correct the problem.

Note that limit expressed here refers to exported data for a single
instrument and collection interval.  The number of timeseries referred
to here corresponds with the number of distinct data points produced
by the instrument over one collection interval.

Under some circumstances, depending on the configured aggregation
temporality, the MeterProvider may be able to recover after these

> limits are reached.

Reviewer 7: This seems to contradict the fact that "metrics for the instrument MUST be dropped and an error reported to the user". Should the drop behavior be recommended and the exception called out explicitly that it shouldn't be dropped when the MeterProvider can recover?

Author: I was referring to a potential implementation detail, but apparently the detail is not clear after my edits. If an SDK is using delta temporality and uses certain implementation techniques (which I'd like to keep out of scope), then the cardinality limit can be evaluated on a per-export basis; the loss of metrics is temporary in this case, whereas with cumulative temporality it seems fruitless to try to recover.

Author: PTAL latest changes.

### Specific cardinality limits

There is only one configurable limit: the number of timeseries per
instrument, with recommended default value 2000.

> SDKs SHOULD NOT support unlimited cardinality.  There is no unset

Reviewer 7: Is there a reason to not make this a requirement? When would it be valid for an implementation to support unlimited cardinality?

Author: An existing stable SDK that supports unlimited cardinality might object to this as a harder requirement. I don't object.

Author: PTAL latest changes.

> SDKs SHOULD NOT support unlimited cardinality.  There is no unset

Reviewer 7: If an SDK doesn't support cardnality limits, it will support unlimited cardinality. Given cardnality limits are recommendations, this statement currently only states the same thing as "MeterProviders SHOULD support being configured with limits". I think the intent is the following though:

```suggestion
SDKs SHOULD NOT support the configuration of unlimited limits.  There is no unset
```

time. Attribute truncation or deletion could affect identity of metric time
series and the topic requires further analysis.

## Cardinality limits

**Status**: [Experimental](../document-status.md)

> MeterProviders SHOULD support being configured with a limit applied to

Reviewer 9: Is this something that should be at the MeterProvider when it looks like it's setting a per-Instrument or per-Aggreagtion limit?

Author: In my prototype w/ the LS metrics SDK ([as detailed below](https://github.com/open-telemetry/opentelemetry-specification/pull/2960#issuecomment-1454203830)), I used an instrument-level limit and a View-level limit. The instrument-level limit is only needed because there's an intermediate data structure where an independent limit makes sense. I excluded this part from this PR, which only proposes (now) a view-level cardinality limit so that views can correct overflow problems.

metrics pipelines from excessive data production in cases when the
number of timeseries produced by application code grows large, which
can happen due to several factors.

Whether because of single attributes having many distinct values or
because of combinatorial expansion among many attributes, these limits
help protect the overall system from individual sources of excessive
metrics instrumentation.

The specific limit, known as `MaxTimeseries` is configured as an
option to the SDK during initialization.  In a metrics pipeline, every
unqiue attribute set maps onto at most one unique
[timeseries](./data-model.md#timeseries-model).  Although the limit
controls the number of timeseries produced by a metrics pipeline, the
SDKs SHOULD apply this limit as early as possible in the metrics
pipeline to prevent buildup of memory.

In this version of the specification, the SDK's `MaxTimeseries`
setting limits all instruments and all instrumentation libraries
equally; future versions of this specification may support finer-grain
configuration of cardinality limits.

Implementations details determine how concurrent API events will be
handled.  To enforce the `MaxTimeseries` limit:

> - An instrument with no configured views SHOULD NOT implement

Reviewer 7: Yeah, I'm a bit confused here as well. The default aggregations will still be used if no view is provided by the user, right? Wouldn't we want to have limits in this situation as well?

> - An instrument with no configured views SHOULD NOT implement

Reviewer 9: ```suggestion
- An instrument that is selected by no views SHOULD NOT implement
  cardinality limits.
```
I was confused by this as well.  Are we trying to say that if the instrument is not selected by a view, and thus won't export, does not need to implement this?  Assuming that it will be a drop aggregation and not store any of these values anyways.

> - An instrument with no configured views SHOULD NOT implement

Author: I replaced this older approach with a new one in which Views configure limits. PTAL.

  cardinality limits.
- An instrument with one or more configured views SHOULD detect that
  the limit was reached before or during the next collection event.

> - An instrument is NOT REQUIRED to be remove stale timeseries (e.g.,

Reviewer 7: ```suggestion
- An instrument is NOT REQUIRED to remove stale timeseries (e.g.,
```

  exclude zero-valued sums from consideration) before evaluating the
  cardinality limit.

> When the limit is reached by an the individual metric instrument,

Reviewer 9: How does this interact with filters in a view?  Should the limit be counted before any attributes are removed or after?  If I have two instruments that filter to the same time series how is the limit calculated?

Author: I replaced this older approach with a new one in which Views configure limits. PTAL.

subsequent new timeseries created for that instrument will have their

> complete attribute set replaced by a single-attribute named `overflow`

Reviewer 7: Should this use the `otel.*` namespace?

Maybe?

```suggestion
complete attribute set replaced by a single-attribute named `otel.metric.instrument.overflow`
```

Author: May I propose `otel.metric.overflow=true`? I'm afraid the extra `instrument` is a little too verbose.

> complete attribute set replaced by a single-attribute named `overflow`

Reviewer 7: This means that when the limit is hit for an instrument the only data that will be sent to an export pipeline is one attribute ("overflow", `true`) and the summed value of all measurements, right?

Author: I replaced this older approach with a new one in which Views configure limits. PTAL.

> with value `true`.

Reviewer 7: ```suggestion
with the boolean value `true`.
```

Author: Fixed.

The implementation is permitted to evaluate the cardinality limit at a
time when it is convenient.  Implementations are NOT REQUIRED to
exactly meet the cardinality limit when it is reached, because that
implies an unwanted degree of synchronization.  Implementations MAY
exceed the limit due to the effects of concurrency.  Implementations
MAY defer the overflow test until the nexts collection cycle.

When cardinality limits are reached, the implementation is REQUIRED

> eventually to stop creating new timeseries for the instrument.

Reviewer 7: This is a pretty vague requirement.

If we don't have hard criteria for when this should happen we should probably just make this a recommendations to the SDK authors to do this as soon as they can to save memory.

Or maybe the requirement is just that the algorithm used needs to ensure there is never a situation where an infinite amount of memory would be used before the limit is applied?

Author: I replaced this older approach with a new one in which Views configure limits. PTAL.

Concerning the synchronous instruments, the implementation of
cardinality limits is NOT REQUIRED to ensure that metric API events
are mapped uniquely onto a single timeseries when the instrument has
reached its cardinality limit.

The SDK SHOULD try to maintain correctness for timeseries that existed
before the overflow event.  However, if this feature requires
substantial additional memory the SDK SHOULD prefer to export an
arbitrary subset of timeseries when the limit is reached.

### Specific cardinality limits

The SDKs that support cardinality limits MUST support the configuration and enforcement of the following limits.

> - The number of timeseries per instrument, with a RECOMMENDED default value 2000.

Reviewer 7: I think from the above section this is per collection cycle, right?

> - The number of timeseries per instrument, with a RECOMMENDED default value 2000.

Author: Yes, I clarified this point. PTAL.

> All SDKs MUST NOT support the configuration or enforcement of any limit not listed above.

Reviewer 7: @jack-berg I thought Java already has other limits, right? Or am I miss-remembering?

Reviewer: No we currently just have a hardcoded limit of 2000 per instrument.

  the background task calls `Collect()` which passes metrics to the push
  exporter.

### MetricReader configurable limits

**Status**: [Experimental](../document-status.md)

MetricReader SHOULD support being configured with limits to be applied
to metrics produced from a single Instrumentation Library Scope.  This
mechanism supports protecting metrics pipelines from excessive data
production, which is usually caused by a problem known as "explosition

> of cardinality", in which the number of timeseries produced by the SDK

Reviewer 2: Considering folks who might not have clear understanding, we've mentioned "cardinality explosion" in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/supplementary-guidelines.md#synchronous-example-cumulative-aggregation-temporality. Maybe we can have a dedicated section in the supplementary guideline doc and link from here, WDYT? (I can send a follow-up PR if folks like the idea).

Author: Sounds good. I added a link from this passage to the section in supplementary guidelines where the term appear for now.

  ranked by cardinality).
- Recommend the user to configure a [View](#view) that filters the
  offending cardinality or drops selected instruments to correct the problem.

Note that limits expressed here refer to exported data for a single
collection cycle.  The term "timeseries" in each of these limits
indicates the number of data points produced over one interval.

Under some circumstances, depending on the configured aggregation
temporality, the SDK may be able to recover after these limits are
reached.  In particular, for Synchronous instruments configured with
Delta aggregation temporality, the SDK SHOULD remove timeseries from
memory when they become stale, making it possible to recover from an
explosion of cardinality from one collection cycle to the next
provided the application stops using excessive cardinality.  However,
for Readers that have Synchronous instruments configured with
Cumulative aggregation temporality, once a limit is reached the
library cannot recover and the SDK SHOULD disable it for the remaining
lifetime of the SDK.

The configurable limits are:

1. The number of timeseries per instrument, with recommended default value 2000.

> 2. The number of timeseries per instrumentation library scope, with recommended default value 2000.

Reviewer: What's the use case for limiting the timeseries for a particular scope? I feel like most of the time users will opt to limit the number of timeseries per instrument because its easier to reason about. It will be easier to implement instrument limits than scope limits since scope limits will require that each instrument not only be aware of its own cardinality but also the other instruments in the scope.

Reviewer 2: I guess similar to tiered throttling - to prevent a certain scope going wild and eating up all the quota?

Author: I was thinking in terms of mental effort for the person configuring these limits. If the goal is to limit the cost, then it's easier to reason about per-library limits than per-instrument limits, since otherwise the cost depends on the number of instruments (which could change). I would argue against provider-level limits because the failure mode for a provider-level limit-exceeded is not very good and there's not much the user can do.

Reviewer: > Iit's easier to reason about per-library limits than per-instrument limits

We're in opinion territory, but I think per-instrument limits are easier to reason about. 

In any case, we need to weigh the benefit of scope level limits against the cost of implementation. I'd prefer to not have to coordinate with a lock across instruments.

> otherwise the cost depends on the number of instruments (which could change).

Maybe in some cases but most of the time the number of instruments will be constant. If you're dynamically creating instruments I imagine most of the time the variable which is changing is better suited as an attribute.

* The `exporter` to use, which is a `MetricExporter` instance.
* The default output `aggregation` (optional), a function of instrument kind.  If not configured, the [default aggregation](#default-aggregation) SHOULD be used.
* The default output `temporality` (optional), a function of instrument kind.  If not configured, the Cumulative temporality SHOULD be used.

> * The default aggregation cardinality limit to use, a function of instrument kind.  If not configured, a default value of 2000 SHOULD be used.

Reviewer 9: Isn't the limit view-scoped?  Why would we need the instrument kind?

Author: I think we're having some kind of terminology mix-up. I don't intend for there to be interaction between instruments. When I refer to a single "view" I mean the one or more individual aggregations that results from evaluating all the View inputs. 

[This terminology is not precise in the specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#view), in any case I would suggest we add the term "aggregation" to the proposed limit, as in `aggregation_cardinality_limit` to indicate that the limit is per instrument-output. I want it to be clear that there's no synchronization required between instruments or between aggregations assigned to the same instrument.

* The `exporter` to use, which is a `MetricExporter` instance.
* The default output `aggregation` (optional), a function of instrument kind.  If not configured, the [default aggregation](#default-aggregation) SHOULD be used.
* The default output `temporality` (optional), a function of instrument kind.  If not configured, the Cumulative temporality SHOULD be used.

> * The default aggregation cardinality limit to use, a function of instrument kind.  If not configured, a default value of 2000 SHOULD be used.

Author: I was trying to follow the existing configuration structure, to simplify the task for SDK authors. Since there are already per-kind defaults, adding a new per-kind default made sense to me. OTOH, I see your point and I would be glad to specify an SDK-wide default without reference to instrument kind.

> * The default aggregation cardinality limit to use, a function of instrument kind.  If not configured, a default value of 2000 SHOULD be used.

Reviewer 3: I don't have a strong opinion that a per-kind default should not exist, so I guess an SDK-wide default could be additive to what you have here in this PR.

I was more considering this from an end user usability perspective. Could we say "SDKs MAY also offer a way to set a global default irrespective of instrument kind. Any limit applied to instrument kind overrides the global default."?

> * The default aggregation cardinality limit to use, a function of instrument kind.  If not configured, a default value of 2000 SHOULD be used.

Reviewer: > specify an SDK-wide default without reference to instrument kind.

For the java implementation I would probably start with this and add the reader / view level options once users ask for it.

> * The default aggregation cardinality limit to use, a function of instrument kind.  If not configured, a default value of 2000 SHOULD be used.

Author: @jack-berg does the "a function of instrument kind" part bother you, or does it make sense? I was thinking about how histograms are more expensive to send than other instruments, to a different limit for histograms than for other points could be useful.

> * The default aggregation cardinality limit to use, a function of instrument kind.  If not configured, a default value of 2000 SHOULD be used.

Reviewer: Although I don't think it will be very popular, I don't think it bothers me. Just because a user has the option of changing the cardinality limit as a function of instrument kind doesn't mean they have to. I don't think it changes to the complexity of the implementation much to offer it.
